src/ray/rpc/server_call.h

```diff
@@ -200,11 +205,8 @@ class ServerCallImpl : public ServerCall {
           // is async and this `ServerCall` might be deleted right after `SendReply`.
           send_reply_success_callback_ = std::move(success);
           send_reply_failure_callback_ = std::move(failure);
-
-          // When the handler is done with the request, tell gRPC to finish this request.
-          // Must send reply at the bottom of this callback, once we invoke this funciton,
-          // this server call might be deleted
-          SendReply(status);
+          boost::asio::post(GetServerCallExecutor(),
+                            [this, status]() { SendReply(status); });
```

Comment:
Can you comment `SendReply ` is thread-safe?

Reply:
No, SendReply is not thread safe. Here at the same time, only one SendReply is going to be called per object.

Reply:
hmm seems SendReply is accessing shared state in this object, would this cause thread safety issue?

Reply:
Yes, but I don't think that's the cause of the test failure
It's because grpc shutdown not taking care of in-flight requests.

But the read and write is also an issue here. Haven't seen anything wrong with it yet.

I'm going to hold the PR for a while. Still targeting 2.2.